src/System.Net.Sockets/src/System/Net/Sockets/SocketTaskExtensions.cs

```diff
@@ -66,69 +101,108 @@ public static Task<int> ReceiveAsync(this Socket socket, ArraySegment<byte> buff
             SocketFlags socketFlags, 
             EndPoint remoteEndPoint)
         {
+            Tuple<Socket, EndPoint> packedArguments = Tuple.Create(socket, remoteEndPoint);
+
             return Task<SocketReceiveFromResult>.Factory.FromAsync(
-                (callback, state) => socket.BeginReceiveFrom(
-                    buffer.Array, 
-                    buffer.Offset, 
-                    buffer.Count, 
-                    socketFlags, 
-                    ref remoteEndPoint, 
-                    callback, 
-                    state),
-                asyncResult => {
-                    int bytesReceived = socket.EndReceiveFrom(asyncResult, ref remoteEndPoint);
-
-                    return new SocketReceiveFromResult() {
+                (targetBuffer, flags, callback, state) =>
+                {
+                    var arguments = (Tuple<Socket, EndPoint>)state;
+                    Socket s = arguments.Item1;
+                    EndPoint e = arguments.Item2;
+
+                    return s.BeginReceiveFrom(
+                        targetBuffer.Array,
+                        targetBuffer.Offset,
+                        targetBuffer.Count,
+                        flags,
+                        ref e,
```

Comment:
Yes, you are right.
About test gap, [seems it's intentionally](https://github.com/dotnet/corefx/blob/3669a883eae0ee87f2db20f88005a26b825f0dc2/src/System.Net.Sockets.Legacy/tests/FunctionalTests/DualModeSocketTest.cs#L1768L1784).
`BeginReceiveFrom` in tests it's a helper extensions which fallback to `ReceiveFromAsync` extension.